### ISSUE TITLE
CI build failing for OSX due to ImportError: cannot import name 'get_all_providers'

### ISSUE THREAD

**Issue by Drulex** [neuropoly/spinalcordtoolbox#2930]
https://travis-ci.org/github/neuropoly/spinalcordtoolbox/jobs/731373490#L922

**Comment by Drulex** [comment]
From slack thread:
> So yesterday I made a commit that broke master. I reverted it today (https://github.com/neuropoly/spinalcordtoolbox/commit/877ad6316d3d94b918d2185f90dcaf8417c1e1fc), but it seems that another bug lurked in silently in the meantime because master is now failing on OSX (https://travis-ci.org/github/neuropoly/spinalcordtoolbox/jobs/731373490#L922). Seems related to an ivadomed import

**Comment by joshuacwnewton** [comment]
Aha. Diffing the two logs for both jobs yields a key difference:

```diff
Successfully installed Keras-2.1.5 PyQt5-sip-4.19.19 PyWavelets-1.1.1 absl-py-0.10.0 aiohttp-3.6.2 
async-timeout-3.0.1 attrs-20.2.0 bids-neuropoly-0.2 bids-validator-1.5.4 bleach-1.5.0 cachetools-4.1.1 
cffi-1.14.3 chardet-3.0.4 click-7.1.2 colored-1.4.2 coverage-5.3 cryptography-3.1.1 cycler-0.10.0 
decorator-4.4.2 dipy-1.2.0 docopt-0.6.2 future-0.18.2 futures-3.1.1 google-auth-1.22.0 
google-auth-oauthlib-0.4.1 grpcio-1.32.0 h5py-2.10.0 html5lib-0.9999999 idna-2.10 idna-ssl-1.1.0 
imageio-2.9.0 importlib-metadata-2.0.0 iniconfig-1.0.1 ivadomed-2.0.2 joblib-0.16.0 kiwisolver-1.2.0 
markdown-3.2.2 matplotlib-3.3.2 multidict-4.7.6 networkx-2.5 nibabel-3.1.1 num2words-0.5.10 oauthlib-3.1.0 
- onnxruntime-1.4.0 
+ onnxruntime-1.5.1
packaging-20.4 pandas-1.1.2 patsy-0.5.1 pillow-7.2.0 pluggy-0.13.1 protobuf-3.13.0 psutil-5.7.2 py-1.9.0 
pyOpenSSL-19.1.0 pyasn1-0.4.8 pyasn1-modules-0.2.8 pybids-0.12.1 pycparser-2.20 pyparsing-2.4.7 
pyqt5-5.11.3 pytest-6.1.0 pytest-cov-2.10.1 python-dateutil-2.8.1 pytz-2020.1 pyyaml-5.3.1 raven-6.10.0 
requests-2.24.0 requests-oauthlib-1.3.0 requirements-parser-0.2.0 rsa-4.6 scikit-image-0.17.2 scikit-learn-0.23.2
 scipy-1.5.2 seaborn-0.11.0 six-1.15.0 sqlalchemy-1.3.19 tensorboard-2.3.0 tensorboard-plugin-wit-1.7.0 
tensorflow-1.5.0 tensorflow-tensorboard-1.5.1 threadpoolctl-2.1.0 tifffile-2020.9.3 toml-0.10.1 torch-1.5.0 
torchvision-0.6.0 tqdm-4.50.0 transforms3d-0.3.1 typing-extensions-3.7.4.3 urllib3-1.25.10 werkzeug-1.0.1 
xlwt-1.3.0 yarl-1.6.0 zipp-3.2.0
```

Considering the error message is:

```
python/envs/venv_sct/lib/python3.6/site-packages/ivadomed/utils.py:8: in <module>
    import onnxruntime
python/envs/venv_sct/lib/python3.6/site-packages/onnxruntime/__init__.py:13: in <module>
    from onnxruntime.capi._pybind_state import get_all_providers, get_available_providers, get_device, set_seed, \
E   ImportError: cannot import name 'get_all_providers'
```

I'm fairly certain this version bump for `onnxruntime` is the cause.

**Comment by jcohenadad** [comment]
great catch @joshuacwnewton !

**Comment by joshuacwnewton** [comment]
I've opened an issue at the `onnxruntime` project to see if they have any insight here. :slightly_smiling_face: 

https://github.com/microsoft/onnxruntime/issues/5344

**Comment by joshuacwnewton** [comment]
After merging #2932, the Mojave (10.14) build has been fixed, but the High Sierra (10.13) and Sierra (10.12) builds are still failing. See: https://travis-ci.org/github/neuropoly/spinalcordtoolbox/builds/732562467

The error from those jobs is roughly the same. 10.13:

```
6.89s$ rvm $brew_ruby do brew bundle --verbose --global
/usr/local/bin/brew tap homebrew/bundle
==> Tapping homebrew/bundle
Cloning into '/usr/local/Homebrew/Library/Taps/homebrew/homebrew-bundle'...
remote: Enumerating objects: 96, done.
remote: Counting objects: 100% (96/96), done.
remote: Compressing objects: 100% (89/89), done.
remote: Total 96 (delta 5), reused 27 (delta 1), pack-reused 0
Receiving objects: 100% (96/96), 50.17 KiB | 3.34 MiB/s, done.
Resolving deltas: 100% (5/5), done.
Tapped (105 files, 265.5KB).
Error: Unknown command: bundle
No Gemfile found, skipping bundle install
```

10.12:

```
5.55s$ rvm $brew_ruby do brew bundle --verbose --global
/usr/local/bin/brew tap homebrew/bundle
==> Tapping homebrew/bundle
Cloning into '/usr/local/Homebrew/Library/Taps/homebrew/homebrew-bundle'...
remote: Enumerating objects: 96, done.
remote: Counting objects: 100% (96/96), done.
remote: Compressing objects: 100% (89/89), done.
remote: Total 96 (delta 5), reused 27 (delta 1), pack-reused 0
Unpacking objects: 100% (96/96), done.
Tapped 0 formulae (197 files, 275.0KB)
Error: Unknown command: bundle
Error: Kernel.exit
No Gemfile found, skipping bundle install
```

I don't have enough familiarity with `brew` to diagnose this quickly, so I'm just reporting for now.

**Comment by joshuacwnewton** [comment]
This seems to be a known problem with Travis's 10.12 and 10.13 images. :sweat_smile: 

https://travis-ci.community/t/macos-build-fails-because-of-homebrew-bundle-unknown-command/7296/19

**Comment by joshuacwnewton** [comment]
The fix to update the outdated Homebrew does work for 10.13, High Sierra! 

However... the Travis build is still failing for 10.12: https://travis-ci.org/github/neuropoly/spinalcordtoolbox/builds/733053157

```
Updating Homebrew
$ rvm $brew_ruby do brew update 1>/dev/null
==> Downloading https://homebrew.bintray.com/bottles-portable-ruby/portable-ruby-2.6.3_2.yosemite.bottle.tar.gz
######################################################################## 100.0%
==> Pouring portable-ruby-2.6.3_2.yosemite.bottle.tar.gz
Warning: You are using macOS 10.12.
We (and Apple) do not provide support for this old version.
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Discourse, Twitter or IRC. You are responsible for resolving any issues you
experience while you are running this old version.

[...]

No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```

So, trying to install `libomp` on 10.12 to fix the original `onnxruntime` issue might be a more involved process due to the age of the OS.

**Comment by Drulex** [comment]
> So, trying to install `libomp` on 10.12 to fix the original `onnxruntime` issue might be a more involved process due to the age of the OS.

Looks like the brew update just hanged (hung?) and never returned. Weird.. I'll restart the job on 10.12 to see if it's a glitch or repeatable.

**Comment by joshuacwnewton** [comment]
I'm hitting a wall trying to fix this issue for 10.12. tl;dr: I'm thinking we should drop support for 10.12, which was echoed before in #2841. See below for a recap and justification.

#### Recap 

* `onnxruntime` updated from 1.4.0 to 1.5.1, and now requires `libomp.dylib`. (See https://github.com/microsoft/onnxruntime/issues/5355)
* On macOS, this is installed through Homebrew (`brew install libomp`).
* Homebrew no longer supports 10.12, and emits this warning:  [`Warning: You are using macOS 10.12. We (and Apple) do not provide support for this old version. You will encounter build failures with some formulae. Please create pull requests instead of asking for help on Homebrew's GitHub, Discourse, Twitter or IRC. You are responsible for resolving any issues you experience while you are running this old version.`](https://travis-ci.org/github/neuropoly/spinalcordtoolbox/builds/734634231#L4694)
* And indeed, we _do_ run into build failures. (Inconsistent, sporadic failures, at that.) Over the past couple of days, some of the cron jobs stall with no output, some jobs begin installing many dependencies before `libomp` can be installed (and the job runs out of time), and some jobs give messy errors. 
* My (admittedly hacky) attempts at resolving these issues (#2945) have gotten me nowhere.

#### Justification

I think we have three options:

1. Keep trying to hack together support for 10.12 under the new `onnxruntime=1.5.1` (but risk spending time getting nowhere).
2. Fix `onnxruntime=1.4.0`. This _technically_ worked for #2931, but doesn't feel good at all, and I'm very hesitant to recommend this.
3. Drop support for 10.12.

I am leaning towards option 3, as poor third-party support for 10.12 has caused us issues in the past. (See #2841/#2919, which is [still giving us headaches](https://travis-ci.org/github/neuropoly/spinalcordtoolbox/jobs/730467853).) If we drop support for 10.12, it kills 2 birds with one stone, because we can drop the numpy hack in our `setup.py` without hassle. (@kousu this might be relevant to you.)

Does anyone have any reservations about dropping support? @Drulex @jcohenadad

**Comment by Drulex** [comment]
I think it's reasonable. Apple dropped support for 10.12 in October 2019.
